### PR TITLE
Checkout Failure handling

### PR DESCRIPTION
**Pull request by hani-ibrahim**
working on: #409 , #414 , #474 

- Checkout Failure is triggered if really the error is checkout failed
- An error is displayed in case this occurred and there were address supplied

### CODE REVIEW

**Comment by hani-ibrahim** [comment]
👍

File: AtlasSDK/AtlasSDKTests/APICreateCheckoutSpec.swift [hunk 1]
```diff
@@ -19,13 +19,13 @@ class APICreateCheckoutSpec: APIClientBaseSpec {
             it("should create checkout from article") {
                 self.waitUntilAPIClientIsConfigured { done, client in
                     let sku = "AD541L009-G11"
-                    client.article(forSKU: sku) { result in
+                    client.article(sku) { result in
```
Comment (bonkey) [inline AtlasSDK/AtlasSDKTests/APICreateCheckoutSpec.swift]: Better get used to first named argument :-P

File: AtlasUI/AtlasUI/BannerErrorViewController.swift [hunk 2]
```diff
@@ -71,14 +71,14 @@ class BannerErrorViewController: UIViewController {
...
-        UIView.animateWithDuration(AnimationDuration.normal.rawValue) { [weak self] in
+        UIView.animateWithDuration(AnimationDuration.fast.rawValue) { [weak self] in
             self?.containerView.transform = CGAffineTransformIdentity
         }
     }
 
     func hideBanner() {
         let bannerHeight = containerView.bounds.height
-        UIView.animateWithDuration(AnimationDuration.normal.rawValue, animations: { [weak self] in
+        UIView.animateWithDuration(AnimationDuration.fast.rawValue, animations: { [weak self] in
```
Comment (bonkey) [inline AtlasUI/AtlasUI/BannerErrorViewController.swift]: Wouldn't you prefer something like:

```
UIView.animate(.fast) {
}
```

and similar shortcuts?

File: AtlasUI/AtlasUI/CheckoutSummaryViewController.swift [hunk 3]
```diff
@@ -12,8 +12,7 @@ class CheckoutSummaryViewController: UIViewController, CheckoutProviderType {
         didSet {
             injectCustomer(from: oldValue)
             viewState = checkoutViewModel.checkoutViewState
-            checkPaymentMethod(oldValue)
-            checkPriceChange(oldValue)
+            checkoutViewModel.validateAgainstOldViewModel(oldValue)
```
Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryViewController.swift]: Older than me? ;)

**Comment by bonkey** [comment]
👍

File: AtlasSDK/AtlasSDKTests/APICreateCheckoutSpec.swift (hunk 1 above)
Comment (hani-ibrahim) [inline AtlasSDK/AtlasSDKTests/APICreateCheckoutSpec.swift]: well, yes
but I found this is the only one or two in the client APIs that has his named argument. so better make it consistent for now.
My be we need to fix them all when we migrate to swift 3

File: AtlasUI/AtlasUI/BannerErrorViewController.swift (hunk 2 above)
Comment (hani-ibrahim) [inline AtlasUI/AtlasUI/BannerErrorViewController.swift]: Yea I would :D 
created a new ticket for it https://github.com/zalando-incubator/atlas-ios/projects/3#card-543513

**Comment by hani-ibrahim** [comment]
👍

File: AtlasSDK/AtlasSDKTests/APIArticleSpec.swift [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import Quick
+import Nimble
+
+@testable import AtlasSDK
+
+class APIArticleSpec: APIClientBaseSpec {
+
+    override func spec() {
+
+        describe("Article") {
```
Comment (bonkey) [inline AtlasSDK/AtlasSDKTests/APIArticleSpec.swift]: Why I do see Quick here?

File: AtlasUI/AtlasUI/Utils/UserMessage.swift [hunk 5]
```diff
@@ -76,6 +76,14 @@ struct UserMessage {
         }
     }
 
+    internal static func displayLoader(block: (() -> Void) -> Void) {
```
Comment (bonkey) [inline AtlasUI/AtlasUI/Utils/UserMessage.swift]: Actually, I didn't care to have this function in a different class, but I was worried about it's "magical" nature itself. But let's keep it this way, we can always _refactor_ ⛏

File: AtlasUI/AtlasUITests/CheckoutSummaryViewControllerSpec.swift [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+class CheckoutSummaryViewControllerSpec: QuickSpec {
+
+    override class func setUp() {
+        super.setUp()
+        try! AtlasMockAPI.startServer() // swiftlint:disable:this force_try
+    }
+
+    override class func tearDown() {
+        super.tearDown()
+        try! AtlasMockAPI.stopServer() // swiftlint:disable:this force_try
```
Comment (bonkey) [inline AtlasUI/AtlasUITests/CheckoutSummaryViewControllerSpec.swift]: You can drop all `// swiftlint:` in tests

**Comment by bonkey** [comment]
👍

**Comment by hani-ibrahim** [comment]
👍

**Pull request merged**